File: frontend/src/lib/permissoes.ts
export type Permissoes = {
  ver_movimentacoes: boolean;
  baixar_movimentacoes: boolean;
  gerir_usuarios: boolean;
  desvincular_maquina: boolean;
  alterar_estoque: boolean;
};

export const PADRAO: Record<string, Permissoes> = {
  ADMIN: {
    ver_movimentacoes: true,
    baixar_movimentacoes: true,
    gerir_usuarios: true,
    desvincular_maquina: true,
    alterar_estoque: true,
  },
  OPERACIONAL: {
    ver_movimentacoes: true,
    baixar_movimentacoes: false,
    gerir_usuarios: false,
    desvincular_maquina: true,
    alterar_estoque: true,
  },
  COMERCIAL: {
    ver_movimentacoes: false,
    baixar_movimentacoes: false,
    gerir_usuarios: false,
    desvincular_maquina: true,
    alterar_estoque: false,
  },
};

export const OPCOES_PERFIL = [
  { value: "OPERACIONAL", label: "Operacional" },
  { value: "COMERCIAL", label: "Comercial" },
  { value: "ADMIN", label: "Admin" },
];

export const ROTULOS_PERMISSAO: { chave: keyof Permissoes; label: string }[] = [
  { chave: "ver_movimentacoes", label: "Ver movimentações" },
  { chave: "baixar_movimentacoes", label: "Baixar TXT da semana" },
  { chave: "gerir_usuarios", label: "Gerir usuários" },
  { chave: "desvincular_maquina", label: "Desvincular máquinas" },
  { chave: "alterar_estoque", label: "Cadastrar, editar e excluir" },
];

export function perfilBase(perfil?: string) {
  const valor = (perfil || "OPERACIONAL").toUpperCase();
  if (valor === "COMUM") return "OPERACIONAL";
  return valor in PADRAO ? valor : "OPERACIONAL";
}

export function rotuloPerfil(perfil?: string) {
  const base = perfilBase(perfil);
  if (base === "ADMIN") return "Admin";
  if (base === "COMERCIAL") return "Comercial";
  return "Operacional";
}

export function permissoesDoPerfil(perfil?: string): Permissoes {
  return { ...PADRAO[perfilBase(perfil)] };
}

File: backend/permissoes.py
"""Permissões efetivas por perfil, com override por usuário."""

CHAVES = (
    "ver_movimentacoes",
    "baixar_movimentacoes",
    "gerir_usuarios",
    "desvincular_maquina",
    "alterar_estoque",
)

_TUDO = {chave: True for chave in CHAVES}

PADRAO = {
    "ADMIN": dict(_TUDO),
    "OPERACIONAL": {
        "ver_movimentacoes": True,
        "baixar_movimentacoes": False,
        "gerir_usuarios": False,
        "desvincular_maquina": True,
        "alterar_estoque": True,
    },
    "COMERCIAL": {
        "ver_movimentacoes": False,
        "baixar_movimentacoes": False,
        "gerir_usuarios": False,
        "desvincular_maquina": True,
        "alterar_estoque": False,
    },
}


def normalizar_perfil(perfil: str | None) -> str:
    valor = (perfil or "OPERACIONAL").strip().upper()
    if valor == "COMUM":
        return "OPERACIONAL"
    if valor in PADRAO:
        return valor
    return "OPERACIONAL"


def efetivas(usuario) -> dict:
    base = dict(PADRAO[normalizar_perfil(getattr(usuario, "perfil", None))])
    salvas = getattr(usuario, "permissoes", None)
    if isinstance(salvas, dict):
        for chave in CHAVES:
            if chave in salvas:
                base[chave] = bool(salvas[chave])
    return base


def pode(usuario, chave: str) -> bool:
    return bool(efetivas(usuario).get(chave))


def gravar_permissoes(perfil: str | None, enviadas) -> dict | None:
    if not isinstance(enviadas, dict):
        return None
    padrao = PADRAO[normalizar_perfil(perfil)]
    diferencas = {}
    for chave in CHAVES:
        if chave not in enviadas:
            continue
        valor = bool(enviadas[chave])
        if valor != padrao[chave]:
            diferencas[chave] = valor
    return diferencas or None
